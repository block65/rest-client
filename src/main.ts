// for checking errors thrown
export { ServiceError } from '../lib/errors.js';

// types needed as peer dep for generated clients
export * from '../lib/types.js';

export { Command } from '../lib/command.js';

// a good standard/basic fetcher factory
export { createIsomorphicNativeFetcher } from './fetchers/isomorphic-native-fetch.js';

// the client
export {
  RestServiceClient,
  type RestServiceClientConfig,
} from '../lib/rest-service-client.js';
