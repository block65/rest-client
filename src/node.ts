export * from './index.js';

export { createNodeFetchRequestMethod } from './request-methods/node-fetch.js';
export { createGotRequestMethod } from './request-methods/node-got.js';
