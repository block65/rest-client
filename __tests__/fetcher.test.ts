import { createServer } from 'node:http';
import { afterAll, beforeAll, describe, expect, test } from '@jest/globals';
import getPort from 'get-port';
import { createIsomorphicFetcher } from '../src/main.js';
import { requestListener } from './server.js';

const server = createServer(requestListener);
const isomorphicFetcher = createIsomorphicFetcher();

describe('Fetcher', () => {
  let base: URL;
  beforeAll(async () => {
    const port = await getPort();
    server.listen(port);
    base = new URL(`http://0.0.0.0:${port}`);
  });

  test('200 OK!', async () => {
    const response = await isomorphicFetcher({
      method: 'get',
      url: new URL('/200', base),
    });

    expect(response).toMatchSnapshot({
      url: expect.any(URL),
    });
  });

  test('204', async () => {
    const response = await isomorphicFetcher({
      method: 'get',
      url: new URL('/204', base),
    });

    expect(response).toMatchSnapshot({
      url: expect.any(URL),
    });
  });

  test('JSON Error', async () => {
    expect(
      await isomorphicFetcher({
        method: 'get',
        url: new URL('/json-error', base),
      }),
    ).toMatchSnapshot({
      url: expect.any(URL),
    });
  });

  test('404', async () => {
    expect(
      await isomorphicFetcher({
        method: 'get',
        url: new URL('/404', base),
      }),
    ).toMatchSnapshot({
      url: expect.any(URL),
    });
  });

  test('Custom options iso fetcher', async () => {
    const customOptionsFetcher = createIsomorphicFetcher({
      headers: {
        'x-fetcher': 'custom',
      },
    });

    expect(
      await customOptionsFetcher({
        method: 'get',
        url: new URL('/my-headers', base),
      }),
    ).toMatchSnapshot({
      url: expect.any(URL),
    });
  });

  test('Custom timeout iso fetcher', async () => {
    const customTimeoutFetcher = createIsomorphicFetcher({
      timeout: 100,
    });

    await expect(
      customTimeoutFetcher({
        method: 'get',
        url: new URL('/unresponsive', base),
      }),
    ).rejects.toMatchSnapshot();
  }, 150);

  afterAll((done) => {
    server.close(done);
  });
});
