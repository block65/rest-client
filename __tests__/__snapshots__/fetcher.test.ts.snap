// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetcher 200 OK! 1`] = `
{
  "body": [
    1,
    2,
    3,
  ],
  "ok": true,
  "status": 200,
  "statusText": "OK",
  "url": Any<URL>,
}
`;

exports[`Fetcher 200 empty content ala AWS lambda function URL 1`] = `
{
  "body": "",
  "ok": true,
  "status": 200,
  "statusText": "OK",
  "url": "http://0.0.0.0:39813/aws-lambda-200",
}
`;

exports[`Fetcher 204 1`] = `
{
  "body": null,
  "ok": true,
  "status": 204,
  "statusText": "No Content",
  "url": Any<URL>,
}
`;

exports[`Fetcher 204 empty content ala AWS lambda function URL 1`] = `
{
  "body": null,
  "ok": true,
  "status": 204,
  "statusText": "No Content",
  "url": "http://0.0.0.0:39813/aws-lambda-204",
}
`;

exports[`Fetcher 404 1`] = `
{
  "body": "<h1>Not Found</h1>",
  "ok": false,
  "status": 404,
  "statusText": "Not Found",
  "url": Any<URL>,
}
`;

exports[`Fetcher Custom options iso fetcher 1`] = `
{
  "body": {
    "accept": "*/*",
    "accept-encoding": "gzip, deflate",
    "accept-language": "*",
    "connection": "keep-alive",
    "host": "redacted",
    "sec-fetch-mode": "cors",
    "user-agent": "node",
    "x-fetcher": "custom",
  },
  "ok": true,
  "status": 200,
  "statusText": "OK",
  "url": Any<URL>,
}
`;

exports[`Fetcher JSON Error 1`] = `
{
  "body": {
    "code": 9,
    "message": "Data should be array",
    "status": "FAILED_PRECONDITION",
  },
  "ok": false,
  "status": 400,
  "statusText": "Bad Request",
  "url": Any<URL>,
}
`;
